wallet/src/de/schildbach/wallet/service/CoinJoinService.kt

```diff
@@ -341,25 +435,41 @@ class CoinJoinMixingService @Inject constructor(
                 MixingCompleteListener { _, statusList ->
                     statusList?.let {
                         for (status in it) {
-                            if (status != PoolStatus.FINISHED) {
-                                coroutineScope.launch(Dispatchers.IO) { updateMixingStatus(MixingStatus.ERROR) }
+                            if (status != PoolStatus.FINISHED && status != PoolStatus.ERR_NOT_ENOUGH_FUNDS && status != PoolStatus.ERR_NO_INPUTS) {
+                                coroutineScope.launch { updateMixingState(MixingStatus.ERROR) }
                                 exception = Exception("Mixing stopped before completion ${status.name}")
+                                return@let
                             }
                         }
                     }
                 }
 
+            val sessionCompleteListener = SessionCompleteListener { _, _, _, _, _, _, _ ->
+                coroutineScope.launch {
+                    updateProgress()
+                }
+            }
+
             mixingFinished.addListener({
                 log.info("Mixing complete.")
                 removeMixingCompleteListener(mixingCompleteListener)
+                removeSessionCompleteListener(sessionCompleteListener)
                 if (mixingFinished.get()) {
-                    coroutineScope.launch(Dispatchers.IO) { updateMixingStatus(MixingStatus.FINISHED) }
+                    coroutineScope.launch {
+                        updateProgress()
+                        updateMixingState(MixingStatus.FINISHED)
+                    }
```

Comment:
I removed `Dispatchers.IO` because it seemed to not use the threading pool created above.  This resulted in missing Context exceptions.